### ISSUE TITLE
Add easy way to iterate over table columns

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#3805]
I often find myself doing:

```
for colname in t.columns:
    column = t[colname]
```

It would be nice to provide a way to iterate over columns, for example:

```
for column in t.itercols():
    # ta-da
```

@taldcroft - what do you think?

Now ideally we'd have used ``t.columns`` for that since at the moment ``t.colnames`` and ``t.columns`` seem to do the same thing, but I don't think we want to break backward-compatibility there.

**Comment by taldcroft** [comment]
Yeah, making `t.columns` an OrderedDict instead of a list subclass with dict-like name access is probably one of my biggest regrets for the `Table` design.

I normally do `for column in t.columns.values():`, but I'd be OK having a short cut for that like `t.itercols()`.  (Since the number of columns is generally small there is no performance penalty in Py2 to using `values()`).

I'm unassigning myself on this since it should be an easy PR once we are agreed on doing it and a method name.  I'm not necessarily advocating it, but for symmetry we might have `t.colvals` which is just a property returning `list(t.columns.values())`.  (or more likely the proper six-ified version).

**Comment by embray** [comment]
What's wrong with `for column in t.columns.values():` ?

**Comment by astrofrog** [comment]
@embray - it's not as obvious for users to discover via tab-completion. Having to go two levels deep is not great. If we had an explicit itercols would make it easy to discover.

**Comment by embray** [comment]
Overall I'd prefer @taldcroft's suggestion of making a sequence-like `Columns` container with keyword getitem support.  It would be a backward-incompatible change, but not an impossible one to make.

**Comment by taldcroft** [comment]
We can name the new sequence-like container `gollums` and deprecate `columns`?

**Comment by embray** [comment]
:+1:

**Comment by astrofrog** [comment]
:+1:

**Comment by embray** [comment]
All seriousness aside, if we made `table.columns` a list-like object, would we then do away with methods like `Table.add_column`, `Table.index_column`, `Table.remove_column`, etc. since they would become superfluous with methods on the `table.columns` list-like object?  Or would we keep that object very simple with few special methods?

**Comment by taldcroft** [comment]
> All seriousness aside,

Does that mean we're still joking?

All kidding aside, in the usual fashion I'm actually pretty hesitant to make an API change like this on `Table`.  This feature (`columns` as ordered dict) has been there since 0.1 was released 3 years ago, and in the `Table` predecessor in `ATPy` that @astrofrog and @ebressert wrote.

As with the `Time.format` to `Time.representation` API change that was discussed, the question we need to ask is if the pain to users is worth the improvement.  For me personally I have lots of code using Table here and there, and I have no idea which of it uses `for colname in t.columns` somewhere.  That would make me really annoyed about upgrading to 1.1, knowing that I need to go a do a grep campaign searching for this idiom.  Some of this code is configured Chandra operations code, so updating these would be more than the usual hassle.

So I would say we made a bit of a design mistake early on and now we live with it and pay the price with an extra method, for instance something we could call `itercols()`.  :smile:

**Comment by pulsestaysconstant** [comment]
I'd like to work on this. Should I name it itercols? Or is there a more preferred name I could use?

**Comment by taldcroft** [comment]
`itercols()` is fine by me.

**Comment by algerdnazgul** [comment]
Hi there! @pulsestaysconstant any pull requests?

**Comment by mwcraig** [comment]
It is addressed in #4394, will see @pulsestaysconstant and remind to clean up/rebase.

**Comment by pulsestaysconstant** [comment]
Really sorry about the delay everyone! I ran into some problems with rebasing a while back and it slipped my mind. I'll get it done by the end of this weekend.
